mpi.h

```diff
@@ -178,407 +180,406 @@ typedef struct MPI_ABI_Datatype * MPI_Datatype;
+    MPI_ORDER_FORTRAN                  = 0xF, /* 15 */
+
+    /* Array Datatype Distribution */
+    MPI_DISTRIBUTE_NONE                = 16,
+    MPI_DISTRIBUTE_BLOCK               = 17,
+    MPI_DISTRIBUTE_CYCLIC              = 18,
+    MPI_DISTRIBUTE_DFLT_DARG           = 19,
+
+    /* Datatype Decoding Combiners */
+    MPI_COMBINER_NAMED                 = 101,
+    MPI_COMBINER_DUP                   = 102,
+    MPI_COMBINER_CONTIGUOUS            = 103,
+    MPI_COMBINER_VECTOR                = 104,
+    MPI_COMBINER_HVECTOR               = 105,
+    MPI_COMBINER_INDEXED               = 106,
+    MPI_COMBINER_HINDEXED              = 107,
+    MPI_COMBINER_INDEXED_BLOCK         = 108,
+    MPI_COMBINER_HINDEXED_BLOCK        = 109,
+    MPI_COMBINER_STRUCT                = 110,
+    MPI_COMBINER_SUBARRAY              = 111,
+    MPI_COMBINER_DARRAY                = 112,
+    MPI_COMBINER_F90_INTEGER           = 113,
+    MPI_COMBINER_F90_REAL              = 114,
+    MPI_COMBINER_F90_COMPLEX           = 115,
+    MPI_COMBINER_RESIZED               = 116,
+    MPI_COMBINER_VALUE_INDEX           = 117,
+
+    /* Fortran Datatype Matching */
+    MPIX_TYPECLASS_LOGICAL             = 191,
+    MPI_TYPECLASS_INTEGER              = 192,
+    MPI_TYPECLASS_REAL                 = 193,
+    MPI_TYPECLASS_COMPLEX              = 194,
+
+    /* Communicator and Group Comparisons */
+    MPI_IDENT                          = 201,
+    MPI_CONGRUENT                      = 202,
+    MPI_SIMILAR                        = 203,
+    MPI_UNEQUAL                        = 204,
+
+    /* Communicator Virtual Topology Types */
+    MPI_CART                           = 211,
+    MPI_GRAPH                          = 212,
+    MPI_DIST_GRAPH                     = 213,
+
+    /* Communicator Split Types */
+    MPI_COMM_TYPE_SHARED               = 221,
+    MPI_COMM_TYPE_HW_UNGUIDED          = 222,
+    MPI_COMM_TYPE_HW_GUIDED            = 223,
+    MPI_COMM_TYPE_RESOURCE_GUIDED      = 224,
+
+    /* Window Lock Types */
+    MPI_LOCK_EXCLUSIVE                 = 301,
+    MPI_LOCK_SHARED                    = 302,
+
+    /* Window Create Flavors */
+    MPI_WIN_FLAVOR_CREATE              = 311,
+    MPI_WIN_FLAVOR_ALLOCATE            = 312,
+    MPI_WIN_FLAVOR_DYNAMIC             = 313,
+    MPI_WIN_FLAVOR_SHARED              = 314,
+
+    /* Window Memory Models */
+    MPI_WIN_UNIFIED                    = 321,
+    MPI_WIN_SEPARATE                   = 322,
+
+    /* File Positioning */
+    MPI_SEEK_SET                       = 401,
+    MPI_SEEK_CUR                       = 402,
+    MPI_SEEK_END                       = 403
 };
 
+/* Other Constants */
+#define MPI_BSEND_OVERHEAD             512 /* MPICH=96, OMPI=128 */
+#define MPI_DISPLACEMENT_CURRENT       ((MPI_Offset)-54278278)
 
-typedef int (MPI_Copy_function)(MPI_Comm, int, void *, void *, void *, int *);
-typedef int (MPI_Delete_function)(MPI_Comm, int, void *, void *);
+/* Predefined Attribute Keys */
+enum {
```

Comment:
@hzhou Jeff bullied me all morning about my changes :rofl: ... He did not like that predef attr keyvals had values that were the same than other unrelated datatype constants. Do you have any strong objection about making them negative with the following values as suggested by Jeff? IMHO, I think your dislike of negative values and its relation to integer overflow do not fully apply here, the use of keyvals on the user side do not involve any arithmetic/increment operations, therefore I don't see a keyval inadvertently overflowing to becoming the same as a predefined key.

There are a few nice things about making predef attr keys negative:
* Plenty of room (the full negative range) for for future MPI standard extension .
* Implementations have the full range of positive values at disposal for user-created keys.
* Negative is predefined, zero is invalid, positive is user-created.

Reply:
If the values are going to be used in arithmetics, then they should be typed as signed integers. Negative numbers is a necessary semantic concept to complete the arithmetic logic. In this usage, small positive integers have the best arithmetic safety.

On the other hand, if the values/variables are not used in arithmetics (including internal code), then its semantics are IDs. The *negative* concept doesn't make sense in an "ID" semantic. People tend to think of an "ID" number as a positive number and it is common to type an "ID" variable as unsigned integers. Thus, negative values for something of an ID are just *weird*. You can, but weird.

That said, it is all my subjective opinion. Using negative numbers for "ID" will work. MPICH has no problem working with either scheme. The cited "nice" points from the implementations' point of view are all kind of moot. :)